src/platform/backends/shared/base_availability_zone_manager.cpp

```diff
@@ -0,0 +1,236 @@
+            }
+
+            mpl::log(mpl::Level::info, category, "using default zones");
+            return default_zones;
+        }
+
+        if (err)
+        {
+            throw AvailabilityZoneManagerDeserializationError{
+                "failed to access {:?}: {}",
+                zones_directory.string(),
+                err.message(),
+            };
+        }
+
+        std::set<std::string> zone_names{};
+        while (zones_dir_iter->hasNext())
+        {
+            const auto& entry = zones_dir_iter->next();
+            if (!entry.is_regular_file())
+                continue;
+            if (entry.path().extension() != ".json")
+                continue;
+
+            mpl::log(mpl::Level::info, category, fmt::format("found AZ file {:?}", entry.path().string()));
+            zone_names.insert(entry.path().stem());
+        }
+        if (zone_names.empty())
+        {
+            mpl::log(mpl::Level::info, category, "no zones found, using defaults");
+            return default_zones;
+        }
+        return zone_names;
+    };
+
+    const auto zone_names = get_zone_names();
+
+    const auto file_type = MP_FILEOPS.status(file_path, err).type();
+    if (err && file_type != fs::file_type::not_found)
+    {
+        throw AvailabilityZoneManagerDeserializationError{
+            "AZ manager file {:?} is not accessible: {}.",
+            file_path.string(),
+            err.message(),
+        };
+    }
+
+    if (file_type == fs::file_type::not_found)
+    {
+        mpl::log(mpl::Level::info,
+                 category,
+                 fmt::format("AZ manager file {:?} not found, using defaults", file_path.string()));
+        automatic_zone = *zone_names.begin();
+        serialize();
+        return;
+    }
+
+    if (file_type != fs::file_type::regular)
+    {
+        throw AvailabilityZoneManagerDeserializationError{
+            "AZ manager file {:?} is not a regular file.",
+            file_path.string(),
+        };
+    }
+
+    mpl::log(mpl::Level::info, category, fmt::format("reading AZ manager from file {:?}", file_path.string()));
+    const auto file = MP_FILEOPS.open_read(file_path);
+    if (file->fail())
+    {
+        throw AvailabilityZoneManagerDeserializationError{
+            "failed to open AZ manager file {:?} for reading: {}",
+            file_path.string(),
+            std::strerror(errno),
+        };
+    }
+
+    const std::string data{std::istreambuf_iterator{*file}, {}};
+    const auto qdata = QString::fromStdString(data).toUtf8();
+    const auto json = QJsonDocument::fromJson(qdata).object();
+
+    if (const auto json_automatic_zone = json[automatic_zone_key].toString().toStdString();
+        zone_names.find(json_automatic_zone) == zone_names.end())
+    {
+        mpl::log(mpl::Level::warning,
+                 category,
+                 fmt::format("automatic zone {:?} not known, using default", json_automatic_zone));
+        automatic_zone = *zone_names.begin();
```

Comment:
Unchecked access & deref on `begin()`, if zone_names is not guaranteed to always have at least 1 element. Still, we shouldn't assume anything and validate our preconditions.

Reply:
Well, it is guaranteed to always have at least one zone name. If none are read from disk, we use the default 3 zone names. If I do one assertion here, do you think that is enough to guarantee the "there is always at least one zone" invariant everywhere downstream from here?